### PR TITLE
Merging Core 2.9.1 including support for cancun/dencun hardfork

### PR DESCRIPTION
**Pull request by mateusz-sekara**
## Solution

### CODE REVIEW

**Comment by makramkd** [comment]
This PR also has the mantle fix, just noting it.

**Review by makramkd — APPROVED** [review]

File: core/chains/evm/client/rpc_client.go [hunk 1]
```diff
@@ -775,21 +776,100 @@ func (r *rpcClient) CallContract(ctx context.Context, msg interface{}, blockNumb
 
 	lggr.Debug("RPC call: evmclient.Client#CallContract")
 	start := time.Now()
+	var hex hexutil.Bytes
 	if http != nil {
-		val, err = http.geth.CallContract(ctx, message, blockNumber)
+		err = http.rpc.CallContext(ctx, &hex, "eth_call", toCallArg(message), toBlockNumArg(blockNumber))
```
Comment (dimkouv) [inline core/chains/evm/client/rpc_client.go:781]: Place `eth_call` in a const? I would expect that it's already defined somewhere.

File: core/chains/evm/client/rpc_client.go [hunk 2]
```diff
@@ -775,21 +776,100 @@ func (r *rpcClient) CallContract(ctx context.Context, msg interface{}, blockNumb
...
+}
+
+func (r *rpcClient) PendingCallContract(ctx context.Context, msg interface{}) (val []byte, err error) {
+	ctx, cancel, ws, http, err := r.makeLiveQueryCtxAndSafeGetClients(ctx)
+	if err != nil {
+		return nil, err
+	}
+	defer cancel()
+	lggr := r.newRqLggr().With("callMsg", msg)
+	message := msg.(ethereum.CallMsg)
```
Comment (dimkouv) [inline core/chains/evm/client/rpc_client.go:806]: check type?
```
message, is := msg.(ethereum.CallMsg)
if !is -> error
```

File: core/chains/evm/client/rpc_client.go [hunk 3]
```diff
@@ -775,21 +776,100 @@ func (r *rpcClient) CallContract(ctx context.Context, msg interface{}, blockNumb
...
+
+	lggr.Debug("RPC call: evmclient.Client#PendingCallContract")
+	start := time.Now()
+	var hex hexutil.Bytes
+	if http != nil {
+		err = http.rpc.CallContext(ctx, &hex, "eth_call", toCallArg(message), "pending")
+		err = r.wrapHTTP(err)
+	} else {
+		err = ws.rpc.CallContext(ctx, &hex, "eth_call", toCallArg(message), "pending")
```
Comment (dimkouv) [inline core/chains/evm/client/rpc_client.go:815]: `pending` const?

File: core/chains/evm/client/rpc_client.go [hunk 4]
```diff
@@ -775,21 +776,100 @@ func (r *rpcClient) CallContract(ctx context.Context, msg interface{}, blockNumb
...
+	if err == nil {
+		val = hex
+	}
+	duration := time.Since(start)
+
+	r.logResult(lggr, err, duration, r.getRPCDomain(), "PendingCallContract",
+		"val", val,
+	)
+
+	return
```
Comment (dimkouv) [inline core/chains/evm/client/rpc_client.go:827]: I think it would be cleaner if you didn't use named return params

File: core/chains/evm/client/rpc_client.go [hunk 5]
```diff
@@ -775,21 +776,100 @@ func (r *rpcClient) CallContract(ctx context.Context, msg interface{}, blockNumb
...
+}
 
+// COPIED FROM go-ethereum/ethclient/gethclient - must be kept up to date!
```
Comment (dimkouv) [inline core/chains/evm/client/rpc_client.go:830]: What will happen if we don't keep this up to date?
I believe this can easily get outdated.
Maybe write some test that uses black magic to somehow check whether this is outdated?

File: core/chains/evm/gas/block_history_estimator.go [hunk 6]
```diff
@@ -858,42 +858,46 @@ func (b *BlockHistoryEstimator) EffectiveGasPrice(block evmtypes.Block, tx evmty
...
 		return nil
 	}
 }
 
+func (b *BlockHistoryEstimator) getEffectiveGasPrice(block evmtypes.Block, tx evmtypes.Transaction) *assets.Wei {
+	if block.BaseFeePerGas == nil || tx.MaxPriorityFeePerGas == nil || tx.MaxFeePerGas == nil {
+		b.logger.Warnw("Got transaction type 0x2 but one of the required EIP1559 fields was missing, falling back to gasPrice", "block", block, "tx", tx)
+		return tx.GasPrice
+	}
+	if tx.GasPrice != nil {
```
Comment (dimkouv) [inline core/chains/evm/gas/block_history_estimator.go:874]: Should this be placed before checking `BaseFeePerGas, MaxPrioFeePerGas is nil` ?

File: core/chains/evm/gas/block_history_estimator.go [hunk 7]
```diff
@@ -858,42 +858,46 @@ func (b *BlockHistoryEstimator) EffectiveGasPrice(block evmtypes.Block, tx evmty
...
+		// Always use the gas price if provided
+		return tx.GasPrice
+	}
+	if tx.MaxFeePerGas.Cmp(block.BaseFeePerGas) < 0 {
+		b.logger.AssumptionViolationw("MaxFeePerGas >= BaseFeePerGas", "block", block, "tx", tx)
+		return nil
+	}
+	if tx.MaxFeePerGas.Cmp(tx.MaxPriorityFeePerGas) < 0 {
+		b.logger.AssumptionViolationw("MaxFeePerGas >= MaxPriorityFeePerGas", "block", block, "tx", tx)
+		return nil
```
Comment (dimkouv) [inline core/chains/evm/gas/block_history_estimator.go:884]: Returning an error would make this function easier to test.
It's hard to test which of this two assumptions were violated without returning an err.

File: core/services/pg/transaction.go [hunk 8]
```diff
@@ -78,7 +78,7 @@ func sqlxTransactionQ(ctx context.Context, db txBeginner, lggr logger.Logger, fn
 				panic(fmt.Sprintf("panic in transaction; aborting rollback that took longer than 10s: %s", p))
 			}
 		} else if err != nil {
-			lggr.Errorf("Error in transaction, rolling back: %s", err)
+			lggr.Warnf("Error in transaction, rolling back: %s", err)
```
Comment (dimkouv) [inline core/services/pg/transaction.go:81]: Why this change?

**Review by dimkouv — APPROVED** [review]
Made the comments cause I thought I'm reviewing a `chainlink` repo PR 😵

**Comment by mateusz-sekara** [comment]
Replaced with https://github.com/smartcontractkit/ccip/pull/582
